specs/core/1_custody-game.md

```diff
@@ -263,45 +275,99 @@ class BeaconBlockBody(Container):
+    for i in range(depth):
+        if index // (2**i) % 2:
+            value = hash(branch[i] + value)
+        else:
+            value = hash(value + branch[i])
+    value = hash(value + mixin.to_bytes(32, "little"))
+    return value == root
 ```
 
 ### `get_crosslink_chunk_count`
 
 ```python
 def get_custody_chunk_count(crosslink: Crosslink) -> int:
     crosslink_length = min(MAX_EPOCHS_PER_CROSSLINK, crosslink.end_epoch - crosslink.start_epoch)
-    chunks_per_epoch = 2 * BYTES_PER_SHARD_BLOCK * SLOTS_PER_EPOCH // BYTES_PER_CUSTODY_CHUNK
-    return crosslink_length * chunks_per_epoch
+    return crosslink_length * CHUNKS_PER_EPOCH
 ```
 
-### `get_bit`
+### `legendre_bit`
+
+Returns the Legendre symbol `(a/q)` normalizes as a bit (i.e. `((a/q) + 1) // 2`). In a production implementation, a well-optimized library (e.g. GMP) should be used for this.
 
 ```python
-def get_bit(serialization: bytes, i: uint64) -> int:
-    """
-    Extract the bit in ``serialization`` at position ``i``.
-    """
-    return (serialization[i // 8] >> (i % 8)) % 2
+def legendre_bit(a: int, q: int) -> int:
+    if a >= q:
+        return legendre_bit(a % q, q)
+    if a == 0:
+        return 0 
+    assert(q > a > 0 and q % 2 == 1)
+    t = 1
+    n = q
+    while a != 0:
+        while a % 2 == 0:
+            a //= 2
+            r = n % 8
+            if r == 3 or r == 5:
+                t = -t
+        a, n = n, a
+        if a % 4 == n % 4 == 3:
+            t = -t
+        a %= n
+    if n == 1:
+        return (t + 1) // 2
+    else:
+        return 0
+```
+
+### ```custody_subchunkify```
+
+Given one proof of custody chunk, returns the proof of custody subchunks of the correct sizes.
+
+```python
+def custody_subchunkify(x: bytes) -> list:
+    subchunks = []
+    start = 0
+    for size in BYTES_PER_CUSTODY_SUBCHUNK:
+        subchunks.append(x[start:start + size])
+        start += size
+    return subchunks
 ```
 
 ### `get_custody_chunk_bit`
 
 ```python
 def get_custody_chunk_bit(key: BLSSignature, chunk: bytes) -> bool:
-    # TODO: Replace with something MPC-friendly, e.g. the Legendre symbol
-    return bool(get_bit(hash(key + chunk), 0))
+    full_G2_element = bls_signature_to_G2(key)
+    s = full_G2_element[0].coeffs
+    bits = [legendre_bit((i + 1) * s[i % 2] + int.from_bytes(subchunk, "little"), BLS12_381_Q)
+            for i, subchunk in enumerate(custody_subchunkify(chunk))]
+
+    return bool(sum(bits) % 2)
 ```
 
 ### `get_chunk_bits_root`
 
 ```python
-def get_chunk_bits_root(chunk_bits: bytes) -> Hash:
-    aggregated_bits = bytearray([0] * 32)
-    for i in range(0, len(chunk_bits), 32):
-        for j in range(32):
-            aggregated_bits[j] ^= chunk_bits[i + j]
-    return hash(aggregated_bits)
+def get_chunk_bits_root(chunk_bits: Bitlist[MAX_CUSTODY_CHUNKS]) -> bit:
+    aggregated_bits = 0
+    for i, b in enumerate(chunk_bits):
+        aggregated_bits += 2**i * b
+    return legendre_bit(aggregated_bits, BLS12_381_Q)
```

Comment:
This function uses bigints (and even unlimited-size ones), which have at times been viewed as a no-no. Is this a problem?

Reply:
True. Since all the computations are mod `q`, 48 bytes/384 bits would be enough, so we could add that size (or 512 bits) as an allowed int size. The Legendre bit computation function obviously needs it as well. I think rewriting this logic using only 256 bit integers would be a very unelegant solution to this problem.

Reply:
Actually, given that the modulus is 381 bits, it needs 768 bit integers to store intermediate (pre-reduction) values. I think the case for Bigints is actually strong here (and I would suggest all implementers use an actual Bignum library like GMP).